internal/app/cache/cache.go

```diff
@@ -17,3 +22,104 @@ type Cache interface {
 	// AddWatch adds the watch to the cache and returns whether a stream is open.
 	AddWatch(key string, req envoy_api_v2.DiscoveryRequest) (bool, error)
 }
+
+type cache struct {
+	cache         *ristretto.Cache
+	expireSeconds time.Duration
+}
+
+type resource struct {
+	resp       *envoy_api_v2.DiscoveryResponse
+	watches    []*envoy_api_v2.DiscoveryRequest
+	streamOpen bool
+}
+
+func newCache(cacheSizeBytes int64, expireSeconds int) (*cache, error) {
+	config := ristretto.Config{
+		NumCounters: 10 * cacheSizeBytes,
+		MaxCost:     cacheSizeBytes,
+		BufferItems: 64,
+	}
+	newCache, err := ristretto.NewCache(&config)
+	if err != nil {
+		return nil, err
+	}
+	return &cache{
+		cache:         newCache,
+		expireSeconds: time.Duration(expireSeconds) * time.Second,
+	}, nil
+}
+
+func (c *cache) Exists(key string) bool {
+	_, found := c.cache.Get(key)
+	return found
+}
+
+func (c *cache) Fetch(key string) (*envoy_api_v2.DiscoveryResponse, error) {
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, fmt.Errorf("No value found for key: %s", key)
+	}
+	resource, ok := value.(resource)
+	if !ok {
+		return nil, fmt.Errorf("Unable to cast cache value to type resource for key: %s", key)
+	}
+	return resource.resp, nil
+}
+
+func (c *cache) SetResponse(key string, resp envoy_api_v2.DiscoveryResponse) ([]*envoy_api_v2.DiscoveryRequest, error) {
```

Comment:
Not something we have to tackle now, but eventually we're going to either need to:
1. Add logic here responsible for directly notifying watches or,
2. Change the interface here to add observers that the response has changed, so that the orchestrator can subsequently notify watches.